packages/core/src/metrics/MetricsService.ts

```diff
@@ -65,22 +66,27 @@ export class MetricsService {
       throw new Error(`Type is invalid: ${type}`);
     }
 
+    if (!this.publisher || !this.defaultMetrics) {
+      return Promise.resolve();
+    }
+
     const payload: MetricsPayload = {
       clientId: this.getClientId(),
       type,
       timestamp: new Date().getTime(),
       data: {}
     };
 
-    metrics.forEach(m => {
-      payload.data[m.identifier] = m.collect();
-    });
+    const metricsPromise = metrics.concat(this.defaultMetrics)
+      .map(m => m.collect().then(data => {
+        payload.data[m.identifier] = data;
+      }));
 
-    this.defaultMetrics.forEach(m => {
-      payload.data[m.identifier] = m.collect();
+    return Promise.all(metricsPromise).then(() => {
+      if (this.publisher) {
```

Comment:
This is already checked at line 69, I will remove it.

Reply:
I see, I read the recommended approach is to encapsulate it in a const `const publisher = this.publisher` to avoid nested checks.